x-pack/filebeat/input/cometd/input.go

```diff
@@ -0,0 +1,196 @@
+// Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+// or more contributor license agreements. Licensed under the Elastic License;
+// you may not use this file except in compliance with the Elastic License.
+
+package cometd
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/elastic/beats/v7/filebeat/channel"
+	"github.com/elastic/beats/v7/filebeat/input"
+	"github.com/elastic/beats/v7/libbeat/beat"
+	"github.com/elastic/beats/v7/libbeat/common"
+	"github.com/elastic/beats/v7/libbeat/common/atomic"
+	"github.com/elastic/beats/v7/libbeat/logp"
+
+	bay "github.com/elastic/bayeux"
+)
+
+const (
+	inputName = "cometd"
+)
+
+// Run starts the input worker then returns. Only the first invocation
+// will ever start the worker.
+func (in *cometdInput) Run() {
+	var err error
+	in.workerOnce.Do(func() {
+		in.workerWg.Add(1)
+		go func() {
+			in.log.Info("Input worker has started.")
+			defer in.log.Info("Input worker has stopped.")
+			defer in.workerWg.Done()
+			defer in.workerCancel()
+			in.b = bay.Bayeux{}
+			in.creds, err = bay.GetSalesforceCredentials(in.authParams)
+			if err != nil {
+				in.log.Error("not able to get access token", err)
+				return
+			}
+			if err := in.run(); err != nil {
+				in.log.Error(err)
+				return
+			}
+		}()
+	})
+}
+
+func (in *cometdInput) run() error {
+	in.out = in.b.Channel(in.out, "-1", *in.creds, in.config.ChannelName)
```

Comment:
Looking at the implementation here https://github.com/elastic/bayeux/blob/3d75fa9e685c3b344ad0d8f08eb5b15fb542145f/bayeux.go#L269-L270 I can see that this makes HTTP requests via `b.call` but does not take a context as an argument. This means that:

1) This operation can't be cancelled if the parent context of the input is cancelled
2) All HTTP operations have no timeout, and will hang forever if there is no response. bayeux is using `http.NewRequest` when it should be using https://pkg.go.dev/net/http#NewRequestWithContext